File: go/iterator_test.go
package membuffers

import (
	"testing"
	"reflect"
)

var uint32Iterator = []struct{
	buf []byte
	scheme []FieldType
	unions [][]FieldType
	expected []uint32
}{
	{
		[]byte{0x0c,0x00,0x00,0x00, 0x13,0x00,0x00,0x00, 0x14,0x00,0x00,0x00, 0x15,0x00,0x00,0x00},
		[]FieldType{TypeUint32Array},
		[][]FieldType{{}},
		[]uint32{0x13,0x14,0x15},
	},
	{
		[]byte{0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33},
		[]FieldType{TypeUint32Array},
		[][]FieldType{{}},
		[]uint32{},
	},
	{
		[]byte{0x07,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33},
		[]FieldType{TypeUint32Array},
		[][]FieldType{{}},
		[]uint32{0x05, 0x00},
	},
}

func TestIteratorUint32(t *testing.T) {
	for tn, tt := range uint32Iterator {
		m := InternalMessage{}
		m.Init(tt.buf, Offset(len(tt.buf)), tt.scheme, tt.unions)
		res := []uint32{}
		for i := m.GetUint32ArrayIterator(0); i.HasNext(); {
			res = append(res, i.NextUint32())
		}
		if !reflect.DeepEqual(tt.expected, res) {
			t.Fatalf("expected %v but got %v in test #%d", tt.expected, res, tn)
		}
	}
}

func TestIteratorUint8(t *testing.T) {
	buf := []byte{0x03,0x00,0x00,0x00, 0x13, 0x14, 0x15}
	scheme := []FieldType{TypeUint8Array}
	unions :=	[][]FieldType{{}}
	expected := []uint8{0x13,0x14,0x15}
	m := InternalMessage{}
	m.Init(buf, Offset(len(buf)), scheme, unions)
	var res []uint8
	for i := m.GetUint8ArrayIterator(0); i.HasNext(); {
		res = append(res, i.NextUint8())
	}
	if !reflect.DeepEqual(expected, res) {
		t.Fatalf("expected %v but got %v", expected, res)
	}
}

func TestIteratorUint16(t *testing.T) {
	buf := []byte{0x06,0x00,0x00,0x00, 0x13,0x00, 0x14,0x00, 0x15,0x00}
	scheme := []FieldType{TypeUint16Array}
	unions :=	[][]FieldType{{}}
	expected := []uint16{0x13,0x14,0x15}
	m := InternalMessage{}
	m.Init(buf, Offset(len(buf)), scheme, unions)
	var res []uint16
	for i := m.GetUint16ArrayIterator(0); i.HasNext(); {
		res = append(res, i.NextUint16())
	}
	if !reflect.DeepEqual(expected, res) {
		t.Fatalf("expected %v but got %v", expected, res)
	}
}

func TestIteratorUint64(t *testing.T) {
	buf := []byte{0x18,0x00,0x00,0x00, 0x13,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x14,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x15,0x00,0x00,0x00,0x00,0x00,0x00,0x00}
	scheme := []FieldType{TypeUint64Array}
	unions :=	[][]FieldType{{}}
	expected := []uint64{0x13,0x14,0x15}
	m := InternalMessage{}
	m.Init(buf, Offset(len(buf)), scheme, unions)
	var res []uint64
	for i := m.GetUint64ArrayIterator(0); i.HasNext(); {
		res = append(res, i.NextUint64())
	}
	if !reflect.DeepEqual(expected, res) {
		t.Fatalf("expected %v but got %v", expected, res)
	}
}

var messageIterator = []struct{
	buf []byte
	scheme []FieldType
	unions [][]FieldType
	expectedSizes []Offset
}{
	{
		[]byte{0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03,0x00, 0x05,0x00,0x00,0x00, 0x01,0x02,0x03,0x04, 0x05,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03},
		[]FieldType{TypeMessageArray},
		[][]FieldType{{}},
		[]Offset{0x03,0x05,0x03},
	},
	{
		[]byte{0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeMessageArray},
		[][]FieldType{{}},
		[]Offset{0},
	},
	{
		[]byte{0x08,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeMessageArray},
		[][]FieldType{{}},
		[]Offset{0},
	},
	{
		[]byte{0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeMessageArray},
		[][]FieldType{{}},
		[]Offset{},
	},
	{
		[]byte{0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44,0x55},
		[]FieldType{TypeMessageArray},
		[][]FieldType{{}},
		[]Offset{4,0},
	},
}

func TestIteratorMessage(t *testing.T) {
	for tn, tt := range messageIterator {
		m := InternalMessage{}
		m.Init(tt.buf, Offset(len(tt.buf)), tt.scheme, tt.unions)
		res := []Offset{}
		for i := m.GetMessageArrayIterator(0); i.HasNext(); {
			_, size := i.NextMessage()
			res = append(res, size)
		}
		if !reflect.DeepEqual(tt.expectedSizes, res) {
			t.Fatalf("expected %v but got %v in test #%d", tt.expectedSizes, res, tn)
		}
	}
}

var bytesIterator = []struct{
	buf []byte
	scheme []FieldType
	unions [][]FieldType
	expected [][]byte
}{
	{
		[]byte{0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03,0x00, 0x05,0x00,0x00,0x00, 0x01,0x02,0x03,0x04, 0x05,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03},
		[]FieldType{TypeBytesArray},
		[][]FieldType{{}},
		[][]byte{{0x01,0x02,0x03},{0x01,0x02,0x03,0x04,0x05},{0x01,0x02,0x03}},
	},
	{
		[]byte{0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeBytesArray},
		[][]FieldType{{}},
		[][]byte{{}},
	},
	{
		[]byte{0x08,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeBytesArray},
		[][]FieldType{{}},
		[][]byte{{}},
	},
	{
		[]byte{0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44},
		[]FieldType{TypeBytesArray},
		[][]FieldType{{}},
		[][]byte{},
	},
	{
		[]byte{0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44,0x55},
		[]FieldType{TypeBytesArray},
		[][]FieldType{{}},
		[][]byte{{0x11,0x22,0x33,0x44},{}},
	},
}

func TestIteratorBytes(t *testing.T) {
	for tn, tt := range bytesIterator {
		m := InternalMessage{}
		m.Init(tt.buf, Offset(len(tt.buf)), tt.scheme, tt.unions)
		res := [][]byte{}
		for i := m.GetBytesArrayIterator(0); i.HasNext(); {
			res = append(res, i.NextBytes())
		}
		if !reflect.DeepEqual(tt.expected, res) {
			t.Fatalf("expected %v but got %v in test #%d", tt.expected, res, tn)
		}
	}
}

func TestIteratorString(t *testing.T) {
	buf := []byte{0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 'a','b','c',0x00, 0x05,0x00,0x00,0x00, 'h','e','l','l', 'o',0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 'd','e','f'}
	scheme := []FieldType{TypeStringArray}
	unions :=	[][]FieldType{{}}
	expected := []string{"abc","hello","def"}
	m := InternalMessage{}
	m.Init(buf, Offset(len(buf)), scheme, unions)
	res := []string{}
	for i := m.GetStringArrayIterator(0); i.HasNext(); {
		res = append(res, i.NextString())
	}
	if !reflect.DeepEqual(expected, res) {
		t.Fatalf("expected %v but got %v", expected, res)
	}
}

File: javascript/iterator.test.js
import './matcher-extensions';
import {FieldTypes} from './types';
import {ch} from './text';
import {InternalMessage} from "./message";

test('TestIteratorUint32', () => {
  const tests = [
    {
      buf: new Uint8Array([0x0c,0x00,0x00,0x00, 0x13,0x00,0x00,0x00, 0x14,0x00,0x00,0x00, 0x15,0x00,0x00,0x00]),
      scheme: [FieldTypes.TypeUint32Array],
      unions: [],
      expected: [0x13,0x14,0x15],
    },
    {
      buf: new Uint8Array([0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33]),
      scheme: [FieldTypes.TypeUint32Array],
      unions: [],
      expected: [],
    },
    {
      buf: new Uint8Array([0x07,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33]),
      scheme: [FieldTypes.TypeUint32Array],
      unions: [],
      expected: [0x05, 0x00],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getUint32ArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});

test('TestIteratorUint8', () => {
  const tests = [
    {
      buf: new Uint8Array([0x03,0x00,0x00,0x00, 0x13, 0x14, 0x15]),
      scheme: [FieldTypes.TypeUint8Array],
      unions: [],
      expected: [0x13,0x14,0x15],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getUint8ArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});

test('TestIteratorUint16', () => {
  const tests = [
    {
      buf: new Uint8Array([0x06,0x00,0x00,0x00, 0x13,0x00, 0x14,0x00, 0x15,0x00]),
      scheme: [FieldTypes.TypeUint16Array],
      unions: [],
      expected: [0x13,0x14,0x15],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getUint16ArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});

test('TestIteratorUint64', () => {
  const tests = [
    {
      buf: new Uint8Array([0x18,0x00,0x00,0x00, 0x13,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x14,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x15,0x00,0x00,0x00,0x00,0x00,0x00,0x00]),
      scheme: [FieldTypes.TypeUint64Array],
      unions: [],
      expected: [BigInt(0x13),BigInt(0x14),BigInt(0x15)],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getUint64ArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});

test('TestIteratorMessage', () => {
  const tests = [
    {
      buf: new Uint8Array([0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03,0x00, 0x05,0x00,0x00,0x00, 0x01,0x02,0x03,0x04, 0x05,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03]),
      scheme: [FieldTypes.TypeMessageArray],
      unions: [],
      expectedSizes: [0x03,0x05,0x03],
    },
    {
      buf: new Uint8Array([0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeMessageArray],
      unions: [],
      expectedSizes: [0],
    },
    {
      buf: new Uint8Array([0x08,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeMessageArray],
      unions: [],
      expectedSizes: [0],
    },
    {
      buf: new Uint8Array([0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeMessageArray],
      unions: [],
      expectedSizes: [],
    },
    {
      buf: new Uint8Array([0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44,0x55]),
      scheme: [FieldTypes.TypeMessageArray],
      unions: [],
      expectedSizes: [4,0],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getMessageArrayIterator(0);
    const res = [];
    for (const [_, size] of iterator) {
      res.push(size);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expectedSizes);
  }
});

test('TestIteratorBytes', () => {
  const tests = [
    {
      buf: new Uint8Array([0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03,0x00, 0x05,0x00,0x00,0x00, 0x01,0x02,0x03,0x04, 0x05,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, 0x01,0x02,0x03]),
      scheme: [FieldTypes.TypeBytesArray],
      unions: [],
      expected: [new Uint8Array([0x01,0x02,0x03]),new Uint8Array([0x01,0x02,0x03,0x04,0x05]),new Uint8Array([0x01,0x02,0x03])],
    },
    {
      buf: new Uint8Array([0x08,0x00,0x00,0x00, 0x88,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeBytesArray],
      unions: [],
      expected: [new Uint8Array()],
    },
    {
      buf: new Uint8Array([0x08,0x00,0x00,0x00, 0x05,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeBytesArray],
      unions: [],
      expected: [new Uint8Array()],
    },
    {
      buf: new Uint8Array([0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44]),
      scheme: [FieldTypes.TypeBytesArray],
      unions: [],
      expected: [],
    },
    {
      buf: new Uint8Array([0x09,0x00,0x00,0x00, 0x04,0x00,0x00,0x00, 0x11,0x22,0x33,0x44,0x55]),
      scheme: [FieldTypes.TypeBytesArray],
      unions: [],
      expected: [new Uint8Array([0x11,0x22,0x33,0x44]),new Uint8Array()],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getBytesArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});

test('TestIteratorString', () => {
  const tests = [
    {
      buf: new Uint8Array([0x1b,0x00,0x00,0x00, 0x03,0x00,0x00,0x00, ch('a'),ch('b'),ch('c'),0x00, 0x05,0x00,0x00,0x00, ch('h'),ch('e'),ch('l'),ch('l'), ch('o'),0x00,0x00,0x00, 0x03,0x00,0x00,0x00, ch('d'),ch('e'),ch('f')]),
      scheme: [FieldTypes.TypeStringArray],
      unions: [],
      expected: ["abc","hello","def"],
    },
  ];

  for (const tt of tests) {
    const m = new InternalMessage(tt.buf, tt.buf.byteLength, tt.scheme, tt.unions);
    const iterator = m.getStringArrayIterator(0);
    const res = [];
    for (const v of iterator) {
      res.push(v);
    }
    // console.log(tt); // uncomment on failure to find out where
    expect(res).toEqual(tt.expected);
  }
});